README.md
# Skeletons

A Flutter package for building custom skeleton widgets to mimic the page's layout while loading.

## Examples


| Items |  | ListView (Default) |
| ------------- | ------------- | ------------- |
| ![items](gifs/items.gif)  | ![paragraphs](gifs/paragraph.gif)  | ![listview_default](gifs/listview_default.gif)  |

| ListView (Custom) | ListView (Complex Cards) | SkeletonTheme  |
| ------------- |  ------------- | ------------- |
| ![listview_custom](gifs/listview_custom.gif)  | ![listview_cards](gifs/cards_example.gif) | ![skeleton_theme](gifs/skeleton_theme.gif)  |

| Light/Dark modes | Right-To-Left  | Custom Shimmer |
| ------------- | ------------- | ------------- |
| ![light_dark_modes](gifs/light_dark_modes.gif)  | ![rtl](gifs/rtl.gif)  | ![custom_shimmer](gifs/custom_shimmer.gif)  |


All examples can be found here [examples](https://github.com/badjio/skeletons/tree/master/example/lib/examples).


## How To Use

Can be used by encapsulating the child widget in a Skeleton widget:

```dart
import 'package:skeletons/skeletons.dart';

Skeleton(
        isLoading: _isLoading,
        skeleton: SkeletonListView(),
        child: Container(child: Center(child: Text("Content"))),
      )
 ```

 or directly:

 ```dart
 Container(
     child: _isLoading 
     ? SkeletonListView() 
     : Container(child: Center(
                        child: Text("Content"))),
 )

 ```

a `SkeletonTheme` can be used to set the default configs for all skeleton descendants in the tree.

```dart
SkeletonTheme(
    // themeMode: ThemeMode.light,
    shimmerGradient: LinearGradient(
        colors: [
          Color(0xFFD8E3E7),
          Color(0xFFC8D5DA),
          Color(0xFFD8E3E7),
        ],
        stops: [
          0.1,
          0.5,
          0.9,
        ],
      ),
      darkShimmerGradient: LinearGradient(
        colors: [
          Color(0xFF222222),
          Color(0xFF242424),
          Color(0xFF2B2B2B),
          Color(0xFF242424),
          Color(0xFF222222),
        ],
        stops: [
          0.0,
          0.2,
          0.5,
          0.8,
          1,
        ],
        begin: Alignment(-2.4, -0.2),
        end: Alignment(2.4, 0.2),
        tileMode: TileMode.clamp,
      ),
        child: MateriaApp(
            ...
        ),
      ),
```


## More Customization:

| ListView (Complex Cards) |
| ------------- |
| ![listview_cards](gifs/cards_example.gif) |

for more complex shapes you can build your skeleton inside a `SkeletonItem` widget:

```dart
ListView.builder(
  physics: NeverScrollableScrollPhysics(),
  itemCount: 5,
  itemBuilder: (context, index) => Padding(
    padding: const EdgeInsets.all(8.0),
    child: Container(
      padding: const EdgeInsets.all(8.0),
      decoration: BoxDecoration(color: Colors.white),
      child: SkeletonItem(
          child: Column(
        children: [
          Row(
            children: [
              SkeletonAvatar(
                style: SkeletonAvatarStyle(
                    shape: BoxShape.circle, width: 50, height: 50),
              ),
              SizedBox(width: 8),
              Expanded(
                child: SkeletonParagraph(
                  style: SkeletonParagraphStyle(
                      lines: 3,
                      spacing: 6,
                      lineStyle: SkeletonLineStyle(
                        randomLength: true,
                        height: 10,
                        borderRadius: BorderRadius.circular(8),
                        minLength: MediaQuery.of(context).size.width / 6,
                        maxLength: MediaQuery.of(context).size.width / 3,
                      )),
                ),
              )
            ],
          ),
          SizedBox(height: 12),
          SkeletonParagraph(
            style: SkeletonParagraphStyle(
                lines: 3,
                spacing: 6,
                lineStyle: SkeletonLineStyle(
                  randomLength: true,
                  height: 10,
                  borderRadius: BorderRadius.circular(8),
                  minLength: MediaQuery.of(context).size.width / 2,
                )),
          ),
          SizedBox(height: 12),
          SkeletonAvatar(
            style: SkeletonAvatarStyle(
              width: double.infinity,
              minHeight: MediaQuery.of(context).size.height / 8,
              maxHeight: MediaQuery.of(context).size.height / 3,
            ),
          ),
          SizedBox(height: 8),
          Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Row(
                children: [
                  SkeletonAvatar(
                      style: SkeletonAvatarStyle(width: 20, height: 20)),
                  SizedBox(width: 8),
                  SkeletonAvatar(
                      style: SkeletonAvatarStyle(width: 20, height: 20)),
                  SizedBox(width: 8),
                  SkeletonAvatar(
                      style: SkeletonAvatarStyle(width: 20, height: 20)),
                ],
              ),
              SkeletonLine(
                style: SkeletonLineStyle(
                    height: 16,
                    width: 64,
                    borderRadius: BorderRadius.circular(8)),
              )
            ],
          )
        ],
      )),
    ),
  ),
);
```

## Issues and feedback

For issues, please report [here](https://github.com/badjio/skeletons/issues). Contributions are welcome.




CHANGELOG.md
## [0.0.1] - 16.07.2021.

* initial release.

## [0.0.2] - 22.07.2021.

* update SkeletonAvatar.
* add complex Cards example.

## [0.0.3] - 16.07.2021.

* update README.

example/lib/examples/theme_modes.dart
import 'package:flutter/material.dart';
import 'package:skeletons/skeletons.dart';

class ThemeModesExamplePage extends StatefulWidget {
  @override
  _ThemeModesExamplePageState createState() => _ThemeModesExamplePageState();
}

class _ThemeModesExamplePageState extends State<ThemeModesExamplePage> {
  bool _isLoading = true;
  ThemeMode _themeMode = ThemeMode.light;

  void _toggleLoading() {
    setState(() {
      _isLoading = !_isLoading;
    });
  }

  void _toggleThemeMode() {
    setState(() {
      _themeMode =
          _themeMode == ThemeMode.light ? ThemeMode.dark : ThemeMode.light;
    });
  }

  @override
  Widget build(BuildContext context) {
    return SkeletonTheme(
      shimmerGradient: LinearGradient(
        colors: [
          Color(0xFFD8E3E7),
          Color(0xFFC8D5DA),
          Color(0xFFD8E3E7),
        ],
        stops: [
          0.1,
          0.5,
          0.9,
        ],
      ),
      darkShimmerGradient: LinearGradient(
        colors: [
          Color(0xFF222222),
          Color(0xFF242424),
          Color(0xFF2B2B2B),
          Color(0xFF242424),
          Color(0xFF222222),
        ],
        stops: [
          0.0,
          0.2,
          0.5,
          0.8,
          1,
        ],
        begin: Alignment(-2.4, -0.2),
        end: Alignment(2.4, 0.2),
        tileMode: TileMode.clamp,
      ),
      child: MaterialApp(
        themeMode: _themeMode,
        theme: _lightTheme,
        darkTheme: _darkTheme,
        home: Scaffold(
            // backgroundColor: Color(0xFFC7E5EE),
            appBar: AppBar(
              title: Text("Theme Modes"),
              leading: IconButton(
                icon: Icon(
                  Icons.chevron_left,
                  size: 30,
                ),
                onPressed: () => Navigator.of(context).pop(),
              ),
            ),
            body: Skeleton(
              duration: Duration(milliseconds: 800),
              isLoading: _isLoading,
              skeleton: _skeletonView(),
              child: _contentView(),
            ),
            floatingActionButtonLocation:
                FloatingActionButtonLocation.centerDocked,
            floatingActionButton: Padding(
              padding: const EdgeInsets.all(8.0),
              child: Row(
                mainAxisAlignment: MainAxisAlignment.spaceBetween,
                children: <Widget>[
                  FloatingActionButton(
                    heroTag: "THEME_MODE_ACTION_BUTTON",
                    onPressed: _toggleThemeMode,
                    child: Icon(_themeMode == ThemeMode.light
                        ? Icons.wb_sunny
                        : Icons.nightlight_round),
                  ),
                  FloatingActionButton(
                    heroTag: "LOADING_ACTION_BUTTON",
                    onPressed: _toggleLoading,
                    child: Icon(
                      _isLoading
                          ? Icons.hourglass_full
                          : Icons.hourglass_bottom,
                    ),
                  )
                ],
              ),
            )),
      ),
    );
  }

  Widget _skeletonView() => SkeletonListView(
      scrollable: true,
      item: SkeletonListTile(
        contentSpacing: 12,
        verticalSpacing: 12,
        padding: EdgeInsets.symmetric(vertical: 12),
        leadingStyle: SkeletonAvatarStyle(
            width: 64, height: 64, borderRadius: BorderRadius.circular(8)),
        titleStyle: SkeletonLineStyle(borderRadius: BorderRadius.circular(4)),
        subtitleStyle: SkeletonLineStyle(
            borderRadius: BorderRadius.circular(4),
            randomLength: true,
            minLength: MediaQuery.of(context).size.width / 3,
            maxLength: MediaQuery.of(context).size.width / 2),
        hasSubtitle: true,
      ));

  Widget _contentView() => ListView.builder(
        padding: EdgeInsets.symmetric(horizontal: 16.0),
        itemBuilder: (context, index) => Padding(
          padding: const EdgeInsets.symmetric(vertical: 12),
          child: Row(
            children: [
              Container(
                width: 64,
                height: 64,
                decoration: BoxDecoration(
                  borderRadius: BorderRadius.circular(8),
                  image: DecorationImage(
                      image: AssetImage("assets/placeholder.png")),
                ),
              ),
              SizedBox(
                width: 12,
              ),
              Expanded(
                  child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(
                    "Lorem ipsum dolor sit amet.",
                    style: Theme.of(context).textTheme.headline6,
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                  ),
                  SizedBox(
                    height: 8,
                  ),
                  Text("Lorem ipsum dolor sit."),
                ],
              ))
            ],
          ),
        ),
      );
}

ThemeData _lightTheme = ThemeData(
  brightness: Brightness.light,
);

ThemeData _darkTheme = ThemeData(
  brightness: Brightness.dark,
);
